python/src/equistore/operations/unique_metadata.py

```diff
@@ -0,0 +1,253 @@
+    gradient_param: Optional[str] = None,
+) -> Labels:
+    """
+    Returns a :py:class:`Labels` object containing the unique metadata across
+    all blocks of the input :py:class:`TensorMap`  ``tensor``. Unique Labels are
+    returned for the specified ``axis`` (either ``"samples"`` or
+    ``"properties"``) and metatdata ``names``.
+
+    Passing ``gradient_param`` as a ``str`` corresponding to a gradient
+    parameter (for instance ``"cell"`` or ``"positions"``) returns the unique
+    indices only for the gradient blocks. Note that gradient blocks by
+    definition have the same properties metadata as their parent
+    :py:class:`TensorBlock`.
+
+    An empty :py:class:`Labels` object is returned if there are no indices in
+    the (gradient) blocks of ``tensor`` corresponding to the specified ``axis``
+    and ``names``. This will have length zero but the names will be the same as
+    passed in ``names``.
+
+    For example, to find the unique ``"structure"`` indices in the ``"samples"``
+    metadata present in a given :py:class:`TensorMap`:
+
+    .. code-block:: python
+
+        unique_structures = unique_metadata(
+            tensor, axis="samples", names=["structure"],
+        )
+
+    To find the unique ``"atom"`` indices in the ``"samples"`` metadata present
+    in the ``"positions"`` gradient blocks of a given :py:class:`TensorMap`:
+
+    .. code-block:: python
+
+        unique_grad_atoms = unique_metadata(
+            tensor, axis="samples", names=["atom"], gradient_param="positions",
+        )
+
+    :param tensor: the :py:class:`TensorMap` to find unique indices for.
+    :param axis: a ``str``, either ``"samples"`` or ``"properties"``,
+        corresponding to the ``axis`` along which the named unique indices
+        should be found.
+    :param names: a ``str``, ``list`` of ``str``, or ``tuple`` of ``str``
+        corresponding to the name(s) of the indices along the specified ``axis``
+        for which the unique values should be found.
+    :param gradient_param: a ``str`` corresponding to the gradient parameter
+        name for the gradient blocks to find the unique indices for. If none
+        (default), the unique indices of the regular :py:class:`TensorBlock`
+        objects will be calculated.
+
+    :return: a sorted :py:class:`Labels` object containing the unique metadata
+        for the blocks of the input ``tensor`` or its gradient blocks for the
+        specified parameter. Each element in the returned :py:class:`Labels`
+        object has len(``names``) entries.
+    """
+    # Parse input args
+    if not isinstance(tensor, TensorMap):
+        raise TypeError("`tensor` must be an equistore `TensorMap`")
+    names = (
+        [names]
+        if isinstance(names, str)
+        else (list(names) if isinstance(names, tuple) else names)
+    )
+    _check_args(tensor, axis, names, gradient_param)
+    # Make a list of the blocks to find unique indices for
+    if gradient_param is None:
```

Comment:
I do not really understand why you want to pass to _unique_from_blocks, a list of blocks and not directly the tensor. 
- Passing the list force you to make copies (I think but I am not sure that what you are doing here in this if..else statement does a copy but maybe not ) of the blocks, which might be a bit big.
- The list makes you treat differently gradients and blocks while you could just pass the whole TensorMap with all the TensorBlocks and have a flag to make you work on the gradient. This would implies maybe more checks which are avoided here 
- Passing the TensorMap would avoid to do again some sanity checks, e.g. line 246-251 you are checking if the blocks have the same properties which would be guaranteed if the input is a TensorMap.
- Passing the list of block gives a bit of flexibility, for example you can pass blocks from different TensorMaps, but this feature is not really implemented since the input of the public functions is a single TensorMap.

This comment, actually, is embedded in a more general thought about the function that maybe need some discussion. To me it is not really clear what are the use cases, and why you want to use that, thus my comments might be a bit generic and not on point.
From the docstring I understand how the function works but not what are the usecases, I understand that the docstring is not the place where to put usage examples, but they would help me to do a better review.

For the rest the code is pretty clear.

Reply:
Thanks for your comments. 

The reason that the base private function `_unique_from_blocks()` takes a list of blocks is so that it works regardless of whether you pass a list of normal TensorBlocks or Gradient TensorBlocks. It also means that it can be the base function for both `unique_metadata()` and `unique_metadata_block()`, as you can pass the function `blocks=tensor.blocks()` or `[block.gradient(gradient_param) for block in tensor.blocks()]` for the former and `blocks=[block]` or `[block.gradient(gradient_param)]` for the latter. In short, the fundamental operation in the module is to extract the metadata form blocks and find the unique values; the private function `_unique_from_blocks` does this, and the public functions just wrap this for the user. 

I suppose, as you say, the function could handle the input type internally, but to me this would be the same code just moved from the public functions to the private function. You say "This would implies maybe more checks which are avoided here" - what kind of checks do you think are missing by doing it this way round?

As for use cases, the most obvious one is for performing splitting of TensorMaps. If I want to use the `split()` function to divide my TensorMap into smaller TensorMaps, I need to know what the unique values of a given metadata index along a given axis is. For instance:

```py
from equistore import io, Labels
from equistore.operations import unique_metadata, split

tensor = io.load("/path/to/tensor/")

# This is the example code in the docstring of `unique_metadata()`
unique_structures = unique_metadata(tensor, axis="samples", names=["structure"])
unique_structures
>>> Labels([(0,), (1,), (2,), (3,), (4,), (5,), (6,), (7,), (8,), (9,)], dtype=[('structure', '<i4')])

# Shuffle these indices
np.random.shuffle(unique_structures)
unique_structures
>>> Labels([(5,), (8,), (2,), (1,), (4,), (3,), (0,), (7,), (6,), (9,)], dtype=[('structure', '<i4')])

# Split into 2 new TensorMaps based on these unique indices
[tensor_1, tensor_2] = split(
    tensor, axis="samples", grouped_idxs=[unique_structures[:5], unique_structures[5:]]
)

# Verify structure indices in the 1st split tensor
unique_metadata(tensor_1, axis="samples", names=["structure"])
>>> Labels([(5,), (8,), (2,), (1,), (4,)], dtype=[('structure', '<i4')])

# Verify structure indices in the 2nd split tensor
unique_metadata(tensor_2, axis="samples", names=["structure"])
>>> Labels([(3,), (0,), (7,), (6,), (9,)], dtype=[('structure', '<i4')])
```

As presented here, this use of `unique_metadata()` and `split()` forms the foundation for a `split_data()` utility (used for train-test(-validation) splits) function in equisolve (see the `split_data` branch). I see `unique_metadata` as returning a characteristic intrinsic enough to a TensorMap that it belongs in equistore, as opposed to equisolve, for example.

As for your comments on the sanity checks in lines 246-251 - I take your point. The `blocks` it checks will only ever be either a list of length 1 or a list of blocks from a TensorMap. As you say, if the properties are guaranteed to be the same for all blocks, for this check I could remove the iteration over blocks here and just check the first block, i.e. `blocks[0]`. Does this work for you?

Reply:
Anyway you answered my main questin so i think everything is pretty good

Reply:
Can we put the example somewhere and the bring the answer you gave @jwa7 into the code? Discussions in the comments and appearing question might come up again if do not clarify this in source and the docs.

Reply:
Sure, I'll add an example using `split()` into the docstring of `unique_metadata()`